File: sdk/go/registration/deadline.go
package registration

// deadlineItem 是最小堆中的一个 UUID 截止项。
type deadlineItem struct {
	uuid     string
	deadline uint64
}

// deadlineQueue 通过最小堆和 UUID 索引提供 O(log N) 更新、删除与到期弹出。
type deadlineQueue struct {
	items   []deadlineItem
	indices map[string]int
}

// newDeadlineQueue 创建最多预分配 capacity 项的空截止队列。
func newDeadlineQueue(capacity int) deadlineQueue {
	return deadlineQueue{
		items:   make([]deadlineItem, 0, capacity),
		indices: make(map[string]int, capacity),
	}
}

// set 插入或更新 uuid 的 deadline，并只沿可能破坏堆序的方向修复。
func (queue *deadlineQueue) set(uuid string, deadline uint64) {
	if index, exists := queue.indices[uuid]; exists {
		previous := queue.items[index].deadline
		queue.items[index].deadline = deadline
		if deadline < previous {
			queue.up(index)
		} else if deadline > previous {
			queue.down(index)
		}
		return
	}
	index := len(queue.items)
	queue.items = append(queue.items, deadlineItem{uuid: uuid, deadline: deadline})
	queue.indices[uuid] = index
	queue.up(index)
}

// remove 删除 uuid 并报告它是否存在；尾项填零以尽早释放字符串引用。
func (queue *deadlineQueue) remove(uuid string) bool {
	index, exists := queue.indices[uuid]
	if !exists {
		return false
	}
	last := len(queue.items) - 1
	delete(queue.indices, uuid)
	if index == last {
		queue.items[last] = deadlineItem{}
		queue.items = queue.items[:last]
		return true
	}
	// 用尾项填洞，再根据新值与父子关系选择向下或向上修复。
	queue.items[index] = queue.items[last]
	queue.items[last] = deadlineItem{}
	queue.items = queue.items[:last]
	queue.indices[queue.items[index].uuid] = index
	if !queue.down(index) {
		queue.up(index)
	}
	return true
}

// expire 在最早截止时间不晚于 now 时删除并返回对应 UUID。
func (queue *deadlineQueue) expire(now uint64) (string, bool) {
	if len(queue.items) == 0 || queue.items[0].deadline > now {
		return "", false
	}
	uuid := queue.items[0].uuid
	queue.remove(uuid)
	return uuid, true
}

// pop 无条件删除并返回当前最早项；空队列返回 false。
func (queue *deadlineQueue) pop() (string, bool) {
	if len(queue.items) == 0 {
		return "", false
	}
	uuid := queue.items[0].uuid
	queue.remove(uuid)
	return uuid, true
}

// next 只查看当前最早截止时间，不修改队列。
func (queue *deadlineQueue) next() (uint64, bool) {
	if len(queue.items) == 0 {
		return 0, false
	}
	return queue.items[0].deadline, true
}

// up 从 index 向父节点修复最小堆顺序。
func (queue *deadlineQueue) up(index int) {
	for index > 0 {
		parent := (index - 1) / 2
		if !queue.less(index, parent) {
			return
		}
		queue.swap(index, parent)
		index = parent
	}
}

// down 从 index 向较小子节点修复最小堆，并报告元素是否移动。
func (queue *deadlineQueue) down(index int) bool {
	start := index
	for {
		left := index*2 + 1
		if left >= len(queue.items) {
			return index != start
		}
		child := left
		right := left + 1
		if right < len(queue.items) && queue.less(right, left) {
			child = right
		}
		if !queue.less(child, index) {
			return index != start
		}
		queue.swap(child, index)
		index = child
	}
}

// less 比较两个堆项；相同截止时间按 UUID 排序，保证行为确定。
func (queue *deadlineQueue) less(left int, right int) bool {
	if queue.items[left].deadline == queue.items[right].deadline {
		return queue.items[left].uuid < queue.items[right].uuid
	}
	return queue.items[left].deadline < queue.items[right].deadline
}

// swap 交换两个堆项并同步维护 UUID 到索引的反向映射。
func (queue *deadlineQueue) swap(left int, right int) {
	queue.items[left], queue.items[right] = queue.items[right], queue.items[left]
	queue.indices[queue.items[left].uuid] = left
	queue.indices[queue.items[right].uuid] = right
}

File: sdk/rust/src/registration/deadline.rs
use std::{cmp::Ordering, collections::HashMap};

#[derive(Clone)]
struct Item {
    uuid: String,
    deadline: u64,
}

#[derive(Clone, Default)]
pub(crate) struct DeadlineQueue {
    items: Vec<Item>,
    indices: HashMap<String, usize>,
}

impl DeadlineQueue {
    /// 创建可预留 `capacity` 个 UUID 的空索引最小堆。
    pub(crate) fn with_capacity(capacity: usize) -> Self {
        Self {
            items: Vec::with_capacity(capacity),
            indices: HashMap::with_capacity(capacity),
        }
    }

    /// 插入或更新 `uuid` 的绝对 `deadline`。
    ///
    /// 已存在项原地调整堆方向，不分配；新项同时写入堆和 O(1) 位置索引。
    pub(crate) fn set(&mut self, uuid: &str, deadline: u64) {
        if let Some(&index) = self.indices.get(uuid) {
            let previous = self.items[index].deadline;
            self.items[index].deadline = deadline;
            match deadline.cmp(&previous) {
                Ordering::Less => self.up(index),
                Ordering::Equal => {}
                Ordering::Greater => {
                    self.down(index);
                }
            }
            return;
        }
        let index = self.items.len();
        self.items.push(Item {
            uuid: uuid.to_owned(),
            deadline,
        });
        self.indices.insert(uuid.to_owned(), index);
        self.up(index);
    }

    /// 删除 `uuid` 并修复堆与位置索引；不存在时返回 false。
    pub(crate) fn remove(&mut self, uuid: &str) -> bool {
        let Some(index) = self.indices.remove(uuid) else {
            return false;
        };
        let last = self.items.len() - 1;
        if index == last {
            self.items.pop();
            return true;
        }
        self.items.swap_remove(index);
        self.indices.insert(self.items[index].uuid.clone(), index);
        if !self.down(index) {
            self.up(index);
        }
        true
    }

    /// 当最早 deadline 小于等于 `now` 时弹出并返回其 UUID；否则返回 `None`。
    pub(crate) fn expire(&mut self, now: u64) -> Option<String> {
        if self.items.first().is_none_or(|item| item.deadline > now) {
            return None;
        }
        let uuid = self.items[0].uuid.clone();
        self.remove(&uuid);
        Some(uuid)
    }

    /// 无条件弹出最早 deadline 的 UUID；空队列返回 `None`。
    pub(crate) fn pop(&mut self) -> Option<String> {
        let uuid = self.items.first()?.uuid.clone();
        self.remove(&uuid);
        Some(uuid)
    }

    /// 返回最早绝对 deadline，但不修改队列。
    pub(crate) fn next(&self) -> Option<u64> {
        self.items.first().map(|item| item.deadline)
    }

    /// 从 `index` 向父节点上浮，恢复 `(deadline, uuid)` 确定顺序的最小堆性质。
    fn up(&mut self, mut index: usize) {
        while index > 0 {
            let parent = (index - 1) / 2;
            if !self.less(index, parent) {
                break;
            }
            self.swap(index, parent);
            index = parent;
        }
    }

    /// 从 `index` 向最小子节点下沉并恢复堆性质；返回该项是否移动。
    fn down(&mut self, mut index: usize) -> bool {
        let start = index;
        loop {
            let left = index * 2 + 1;
            if left >= self.items.len() {
                return index != start;
            }
            let right = left + 1;
            let child = if right < self.items.len() && self.less(right, left) { right } else { left };
            if !self.less(child, index) {
                return index != start;
            }
            self.swap(index, child);
            index = child;
        }
    }

    /// 比较 `left` 与 `right` 堆项；deadline 相等时用 UUID 提供确定顺序。
    fn less(&self, left: usize, right: usize) -> bool {
        (self.items[left].deadline, &self.items[left].uuid) < (self.items[right].deadline, &self.items[right].uuid)
    }

    /// 交换两个堆位置，并同步更新两个 UUID 的位置索引。
    fn swap(&mut self, left: usize, right: usize) {
        self.items.swap(left, right);
        self.indices.insert(self.items[left].uuid.clone(), left);
        self.indices.insert(self.items[right].uuid.clone(), right);
    }
}

#[cfg(test)]
#[path = "../../tests/internal/registration/deadline.rs"]
mod tests;
